pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_primary_key.go

```diff
@@ -0,0 +1,797 @@
+table with TTL "%s" are not permitted`, ns.Name))
+	// Panic if there is any inbound/outbound FK constraints.
+	_, _, inboundFKElem := scpb.FindForeignKeyConstraint(b.BackReferences(tableID))
+	if inboundFKElem != nil {
+		panic(hasFKConstraintError)
+	}
+	scpb.ForEachForeignKeyConstraint(b.QueryByID(tableID), func(
+		current scpb.Status, target scpb.TargetStatus, e *scpb.ForeignKeyConstraint,
+	) {
+		panic(hasFKConstraintError)
+	})
+}
+
+func mustRetrievePrimaryIndexElement(b BuildCtx, tableID catid.DescID) (res *scpb.PrimaryIndex) {
+	scpb.ForEachPrimaryIndex(b.QueryByID(tableID), func(
+		current scpb.Status, target scpb.TargetStatus, e *scpb.PrimaryIndex,
+	) {
+		// TODO (xiang): for now, we assume there is one primary index, which
+		//               will not be true when there are DDL stmts (e.g.
+		//               ADD/DROP COLUMN) before this `ALTER PRIMARY KEY`.
+		if current == scpb.Status_PUBLIC {
+			res = e
+		}
+	})
+	if res == nil {
+		panic(fmt.Sprintf("programming error: resovling table %v does not give a PrimaryIndex element", tableID))
+	}
+	return res
+}
+
+func mustRetrievePrimaryIndexNameElem(
+	b BuildCtx, tableID catid.DescID, primaryIndexID catid.IndexID,
+) (res *scpb.IndexName) {
+	scpb.ForEachIndexName(b.QueryByID(tableID), func(_ scpb.Status, _ scpb.TargetStatus, e *scpb.IndexName) {
+		if e.IndexID == primaryIndexID {
+			res = e
+		}
+	})
+	if res == nil {
+		panic(fmt.Sprintf("programming error: resolving table %v does not give a PrimaryIndex"+
+			" name element", tableID))
+	}
+	return res
+}
+
+func mustRetrieveColumnElem(
+	b BuildCtx, tableID catid.DescID, columnID catid.ColumnID,
+) (column *scpb.Column) {
+	scpb.ForEachColumn(b.QueryByID(tableID), func(current scpb.Status, target scpb.TargetStatus, e *scpb.Column) {
+		if e.ColumnID == columnID {
+			column = e
+		}
+	})
+	if column == nil {
+		panic(fmt.Sprintf("programming error: cannot find a Column element for column ID %v", columnID))
+	}
+	return column
+}
+
+func mustRetrieveColumnNameElem(
+	b BuildCtx, tableID catid.DescID, columnID catid.ColumnID,
+) (columnName *scpb.ColumnName) {
+	scpb.ForEachColumnName(b.QueryByID(tableID), func(current scpb.Status, target scpb.TargetStatus, e *scpb.ColumnName) {
+		if e.ColumnID == columnID {
+			columnName = e
+		}
+	})
+	if columnName == nil {
+		panic(fmt.Sprintf("programming error: cannot find a ColumnName element for column ID %v", columnID))
+	}
+	return columnName
+}
+
+func mustRetrieveColumnTypeElem(
+	b BuildCtx, tableID catid.DescID, columnID catid.ColumnID,
+) (columnType *scpb.ColumnType) {
+	scpb.ForEachColumnType(b.QueryByID(tableID), func(current scpb.Status, target scpb.TargetStatus, e *scpb.ColumnType) {
+		if e.ColumnID == columnID {
+			columnType = e
+		}
+	})
+	if columnType == nil {
+		panic(fmt.Sprintf("programming error: cannot find a ColumnType element for column ID %v", columnID))
+	}
+	return columnType
+}
+
+func mustRetrieveIndexElement(
```

Comment:
One day before too long I'm going to come back and rework all of this to use rel and to be efficient. So many linear lookup loops.